internal/llrp/device_service.go

```diff
@@ -0,0 +1,294 @@
+//
+// Copyright (C) 2020 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package llrp
+
+import (
+	"bytes"
+	"encoding/json"
+	"github.com/pkg/errors"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strconv"
+)
+
+// These are the names of deviceResource and deviceCommands
+// this expects the devices are using in their registered deviceProfiles.
+const (
+	basePath = "/api/v1/device/name/" // paths are {base/}{device}{/target}
+
+	capDevCmd    = "/capabilities"
+	configDevCmd = "/config"
+	addCmd       = "/roSpec"
+	enableCmd    = "/enableROSpec"
+	disableCmd   = "/disableROSpec"
+	stopCmd      = "/stopROSpec"
+	startCmd     = "/startROSpec"
+	deleteCmd    = "/deleteROSpec"
+
+	enableImpinjCmd = "/enableImpinjExt"
+
+	capReadingName = "ReaderCapabilities"
+
+	maxBody = 100 * 1024
+)
+
+// DSClient is a client to interact with the LLRP Device Service.
+type DSClient struct {
+	baseURL    string
+	httpClient *http.Client
+}
+
+// NewDSClient returns a DSClient reachable at the given host URL,
+// using the given http Client, which of course may be the default.
+func NewDSClient(host *url.URL, c *http.Client) DSClient {
+	base := url.URL{
+		Scheme: host.Scheme,
+		Opaque: host.Opaque,
+		User:   host.User,
+		Host:   host.Host,
+		Path:   basePath,
+	}
+
+	return DSClient{
+		baseURL:    base.String(),
+		httpClient: c,
+	}
+}
+
+// GetDevices return a list of device names known to the EdgeX Metadata service.
+func GetDevices(metadataDevicesURL string, client *http.Client) ([]string, error) {
+	req, err := http.NewRequest(http.MethodGet, metadataDevicesURL, nil)
+	if err != nil {
+		return nil, err
+	}
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, errors.Errorf("GET returned unexpected status: %d", resp.StatusCode)
+	}
+
+	respBody, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	ds := &[]struct{ Name string }{}
+	if err := json.Unmarshal(respBody, ds); err != nil {
+		return nil, errors.Wrap(err, "failed to parse EdgeX device list")
+	}
+
+	deviceList := make([]string, len(*ds))
+	for i, dev := range *ds {
```

Comment:
This should filter out devices not from the LLRP device service. i.e. `dev.Service != "edgex-device-rfid-llrp"`